File: internal/host/resources.go
package host

// The API behind the three tabs a running platform opens on: tasks, datasets
// and labelsets.
//
// These are what makes an export a platform rather than an appliance. Nothing
// here is derived from the pipeline — the pipeline says this platform can run
// annotation tasks, and the people using it decide which tasks, over which
// documents, with which labels, done by whom.

import (
	"encoding/json"
	"net/http"

	"github.com/MaastrichtU-BISS/legal-blocks/internal/db"
)

func (s *server) resourceRoutes(mux *http.ServeMux) {
	mux.HandleFunc("/api/labelsets", s.handleLabelsets)
	mux.HandleFunc("/api/datasets", s.handleDatasets)
	mux.HandleFunc("/api/tasks", s.handleTasks)
}

// handleLabelsets lists or creates labelsets.
func (s *server) handleLabelsets(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		sets, err := s.db.Labelsets(r.Context())
		if err != nil {
			writeError(w, http.StatusInternalServerError, "%v", err)
			return
		}
		writeJSON(w, http.StatusOK, sets)

	case http.MethodPost:
		var body struct {
			Name   string     `json:"name"`
			Desc   string     `json:"desc"`
			Labels []db.Label `json:"labels"`
		}
		if !decodeBody(w, r, &body) {
			return
		}
		owner, err := s.owner(r)
		if err != nil {
			writeError(w, http.StatusInternalServerError, "%v", err)
			return
		}
		id, err := s.db.CreateLabelset(r.Context(), owner, body.Name, body.Desc, body.Labels)
		if err != nil {
			// These failures are things the person filling in the form can
			// fix, so they are 400s carrying the reason rather than 500s.
			writeError(w, http.StatusBadRequest, "%v", err)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"id": id})

	default:
		writeError(w, http.StatusMethodNotAllowed, "use GET or POST")
	}
}

// handleDatasets lists or creates datasets.
//
// A dataset is created with its documents and is not added to afterwards. That
// is a deliberate limit for now: a task's assignments are made from the
// documents present when it was created, so growing a dataset would leave
// existing tasks covering part of it and silently disagree with what the
// screen says they cover.
func (s *server) handleDatasets(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		sets, err := s.db.Datasets(r.Context())
		if err != nil {
			writeError(w, http.StatusInternalServerError, "%v", err)
			return
		}
		writeJSON(w, http.StatusOK, sets)

	case http.MethodPost:
		var body struct {
			Name      string        `json:"name"`
			Desc      string        `json:"desc"`
			Documents []db.Document `json:"documents"`
		}
		if !decodeBody(w, r, &body) {
			return
		}
		if len(body.Documents) == 0 {
			writeError(w, http.StatusBadRequest, "a dataset needs at least one document")
			return
		}
		owner, err := s.owner(r)
		if err != nil {
			writeError(w, http.StatusInternalServerError, "%v", err)
			return
		}
		id, err := s.db.CreateDataset(r.Context(), owner, body.Name, body.Desc, body.Documents)
		if err != nil {
			writeError(w, http.StatusBadRequest, "%v", err)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"id": id})

	default:
		writeError(w, http.StatusMethodNotAllowed, "use GET or POST")
	}
}

// handleTasks lists or creates tasks.
func (s *server) handleTasks(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		tasks, err := s.db.Tasks(r.Context())
		if err != nil {
			writeError(w, http.StatusInternalServerError, "%v", err)
			return
		}
		writeJSON(w, http.StatusOK, tasks)

	case http.MethodPost:
		var spec db.TaskSpec
		if !decodeBody(w, r, &spec) {
			return
		}
		owner, err := s.owner(r)
		if err != nil {
			writeError(w, http.StatusInternalServerError, "%v", err)
			return
		}
		id, err := s.db.CreateTask(r.Context(), owner, spec)
		if err != nil {
			// "that dataset has no documents yet", "a task needs at least one
			// annotator's email address" — all answerable by the person
			// filling in the form.
			writeError(w, http.StatusBadRequest, "%v", err)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"id": id})

	default:
		writeError(w, http.StatusMethodNotAllowed, "use GET or POST")
	}
}

// decodeBody reads a JSON request body, reporting a failure to the client.
func decodeBody(w http.ResponseWriter, r *http.Request, into any) bool {
	if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, maxBodyBytes)).Decode(into); err != nil {
		writeError(w, http.StatusBadRequest, "could not read the request: %v", err)
		return false
	}
	return true
}

File: layers/base/app/api/resources.ts
// The three things a running platform lets people make.

import type { CorpusDocument } from "../types";
import { json, post } from "./http";

export interface Label {
  name: string;
  color?: string;
}

export interface LabelsetSummary {
  id: number;
  name: string;
  desc: string;
  labels: Label[];
  task_count: number;
}

export interface DatasetSummary {
  id: number;
  name: string;
  desc: string;
  documents: number;
  task_count: number;
}

export interface TaskSummary {
  id: number;
  name: string;
  desc: string;
  annotation_level: string;
  dataset_id: number;
  dataset_name: string;
  labelset_id: number;
  labelset_name: string;
  annotators: User[];
  documents: number;
  done: number;
  total: number;
}

export interface TaskSpec {
  name: string;
  desc?: string;
  dataset_id: number;
  labelset_id: number;
  annotation_level: string;
  ann_guidelines?: string;
  /** Email addresses. They need not have used this platform before. */
  annotators: string[];
}

export async function getLabelsets(): Promise<LabelsetSummary[]> {
  return json(await fetch("/api/labelsets"), "loading labelsets");
}

export async function createLabelset(
  name: string,
  desc: string,
  labels: Label[],
): Promise<number> {
  const out = await post<{ id: number }>(
    "/api/labelsets",
    { name, desc, labels },
    "saving the labelset",
  );
  return out.id;
}

export async function getDatasets(): Promise<DatasetSummary[]> {
  return json(await fetch("/api/datasets"), "loading datasets");
}

export async function getTasks(): Promise<TaskSummary[]> {
  return json(await fetch("/api/tasks"), "loading tasks");
}

export async function createTask(spec: TaskSpec): Promise<number> {
  const out = await post<{ id: number }>("/api/tasks", spec, "creating the task");
  return out.id;
}
